DOC: Update install page in Sphinx docs

bjnath opened the pull request:
Removed the install page (user/install.rst) from Sphinx,
since we've put the install instructions in Hugo. install.rst
was inadequate in any case.

The docs front page now leads with "What is NumPy"?

The "Troubleshooting InstallError"
(user/troubleshooting-importerror.rst) page remains, but
since Hugo has the install instructions, it's
linked to from the Hugo install page (via numpy.org PR #360).

bjnath committed:
DOC: Revise glossary page

Entries added, explanations revised,
Python-only entries now point to Python glossary.

bjnath committed:
DOC: Wording improvement in #16996

More clearly describe `shares_memory` as being
not always *feasible* rather than *possible*

bjnath committed:
DOC: Incorporate @BvB93's suggestions in PR #16996

Also reworded `stride` entry for clarity, moved
'axis' entry to correct alphabetical position.

Also added anchors to reference/arrays.indexing for glossary references
to point to.

bjnath committed:
DOC: Avoid unwanted doctest on PR #14979

Failing CI because it's running doctest on the example strings.

Also, deleting glossary.py, source file of the original glossary.

bjnath committed:
DOC: 2nd try: Avoid unwanted doctest on PR #14979

bjnath committed:
DOC: 3rd try: Avoid unwanted doctest on PR #14979

bjnath committed:
DOC: 4th try: Avoid unwanted doctest on PR #14979

bjnath committed:
DOC: 5th try: Avoid unwanted doctest on PR #14979

bjnath committed:
DOC: 6th try: Fix doctest on PR #16996

bjnath committed:
DOC: Fix typos in #16996

strctured -> structured, caught by @WarrenWeckesser.

Fixed missing language in `view` entry.

bjnath committed:
DOC: Fix repeated word in PR #16996

Also adding <BLANKLINE>s to some examples for clarity.

bjnath committed:
DOC: Fix example formatting in PR #16996

rossbar committed:
DOC: Use intersphinx links.

rossbar committed:
DOC: intersphinx standalone glossary entries

rossbar committed:
DOC: rm tuple from glossary

Commit:
ENH,WIP: Added type hints to `np.core._asarray`

Commit:
ENH: Use literals where possible

seberg committed:
MAINT: Simplify ufunc pickling

This also allows at least in principle numba dynamically
generated ufuncs to be pickled (with some hacking), see:

https://github.com/dask/distributed/issues/3450

If the name of the ufunc is set to a qualname, using this method,
pickle should be able to unpickle the ufunc correctly.
We may want to allow setting the module and qualname explicitly
on the ufunc object to remove the need for the custom pickler
completely.

seberg committed:
MAINT: (dtype-transfer) make copyswapn and legacy cast wrapper method

This is just a small preparation to make the code slightly more
modular, since I definitely will need to retain the wrapping for
legacy dtype `dtype->f->castfuncs` and the `dtype->f->copyswapn`,
but want more freedom to see whether I can split out the specialized
casts we use internally (since NumPy has those cast functions, but
normally never uses them).

Splitting these two out should make it easier to write a function
to support current "legacy" dtype cast functions without worrying
as much about the casts used internally for NumPy dtypes.

Commit:
ENH: Add annotations for non-magic ndarray methods

Commit:
MAINT: Moved a number of methods from `_ArrayOrScalarCommon` to `ndarray`

A number of  methods require a >= 1D array or are completelly absent from `generic`.
Move them to `ndarray`.

seberg committed:
BUG: Relax registration of user dtypes with `NPY_ITEM_REFCOUNT`

This also includes `NPY_ITEM_IS_POINTER`.  A previous changed
raised an error when this happened, but some downstream libraries
use it to create a custom dtype with a single object field.

It seems acceptable to create such a dtype if (and only if) that
dtype hardcodes names and fields at creation time, so this change
allows that, but otherwise keeps the error intact.

This should work fine, although some care may be required

seberg committed:
ENH: Make dtype registration time error messages more clear

seberg committed:
MAINT: Relax check that a correct type is already set

This allows running `xpress` on current master. `xpress` copies the
type from `np.dtype(object)` rather than using `&PyArrayDescr_Type`.
That seems overall fine, we can just replace it.
The only reason for this check is to ensure that the user does not
override our updated `type(user_dtype)` and thus corrupting it.

mattip committed:
adjust whitespace to minimize the diff

mattip committed:
DOC: manually merge some changes done in the other PRs

mattip committed:
Merge branch 'master' into revised-glossary

Commit:
TST: Added tests for `np.core._asarray`

Commit:
TST: Consolidate all array construction tests

Commit:
ENH: Update the `np.__all__`

Commit:
MAINT: Minor fixes; removed unused imports

Commit:
ENH: Add an alias for `Union[_Requirements, _E]`

Addresses https://github.com/numpy/numpy/pull/17326#discussion_r489322146

Commit:
ENH: Add annotations for `generic` and `ndarray` bitwise operations

Commit:
MAINT: Moved a few constants back to `__init__.pyi`; cleaned up their names

seberg committed:
TST: Add test for creating an xpress like user-dtype

This includes a few error paths. Note that the test creates
a new dtype ever time it is run, and dtypes in NumPy are persistent
so that this leaks references.

bjnath committed:
DOC: Updated for reviewer comments

Also added "contiguous," which is referenced in docs but hadn't been
defined.

bjnath committed:
DOC: 'Along an axis' rewrite and other reviewer changes

bjnath committed:
DOC: Replace linked headings with explicit 'See'

Commit:
TST: Added new `ndarray`/`generic` typing tests

bjnath committed:
DOC: Fix minor typos

Commit:
DEV: add pr prefix labeler and prefix mapping

bjnath committed:
DOC: How-to guide for drafting how-tos

Migrated from numpy-tutorials

bjnath committed:
DOC: How-to guide for drafting how-tos

Migrated from numpy-tutorials

bjnath committed:
Merge branch 'how-to-guide' of github.com:bjnath/numpy-1 into how-to-guide

# Conflicts:
#	doc/source/user/how-to-how-to.rst

bjnath committed:
DOC: Add how-to index entry for how-to guide PR #17352

bjnath committed:
DOC: Revise #17352 with @mattip suggestions

bjnath committed:
DOC: Remove install instructions from Sphinx docs

Removed the install page (user/install.rst) from Sphinx,
since we've put the install instructions in Hugo. install.rst
was inadequate in any case.

The docs front page now leads with "What is NumPy"?

The "Troubleshooting InstallError"
(troubleshooting-importerror.rst) page remains, but
Hugo has the install instructions it's
 is linked to from the Hugo install page (via numpy.org PR #360).

melissawm committed:
Merge pull request #17352 from bjnath/how-to-guide

DOC: Guide to writing how-tos

Commit:
MAINT: Replaced `__all__` with `import x as x`

Objects defined in an external file will be picked up if they are either specified in `__all__` or imported as `import x as x`.
The latter notation is more compact and is less prone to merge conflicts.

bjnath committed:
DOC: History cleanup for #17353

bjnath committed:
DOC: Update pickle section in PR #17353

WarrenWeckesser committed:
DOC: distutils: Remove an obsolete paragraph.

rgommers committed:
Merge pull request #17369 from WarrenWeckesser/remove-sentence

DOC: distutils: Remove an obsolete paragraph.

eric-wieser commented on doc/source/user/troubleshooting-importerror.rst:
```diff
@@ -1,3 +1,5 @@
+:orphan:
```
Is there somewhere we can link to this so its more discoverable?

rossbar replied:
I'd prefer to leave this where it is, i.e. keep the `Setting Up` section on the front page of the docs. It's an important article and should be discoverable.

rgommers commented:
+1 for removing duplicate info and linking to `numpy.org`.

Would be good to keep https://numpy.org/devdocs/user/building.html linked from somewhere, not sure that's the case now?

bjnath commented:
> its more discoverable

The best place to find it would surely be in the install instructions, which are now in the Hugo pages.  The install page there will link to this page when [numpy.org  PR 360](https://github.com/numpy/numpy.org/pull/360) is merged.

bjnath commented:
@rgommers That page is referenced in [Setting up and using your development environment](https://numpy.org/devdocs/dev/development_environment.html), but it makes sense to add a link on the Hugo install page. I added it to [numpy.org PR 360](https://github.com/numpy/numpy.org/pull/360).

eric-wieser commented on doc/source/user/install.rst:
```diff
@@ -1,10 +0,0 @@
-https://scipy.org/install.html for links to available options.
-
-For instructions on building for source package, see
-:doc:`building`. This information is useful mainly for advanced users.
```
In case users are still linking to this page, should we add a link from here to the Hugo page?

bjnath replied:
@eric-wieser I will do this unless somebody argues otherwise. The reason I've hesitated is that the page might turn up in search and draw attention away from the real installation page. No lasting harm; users will get the Sphinx page and click through to the Hugo page. The issue might not even come up; engines seem to downrank pages that are only a link.

bjnath committed:
DOC: Incorporate blob suggestions in PR #17353

Commit:
TST: Fixed an `int`-related failure on 32-bit systems

Commit:
TST: Attempt #2 at fixing the `int`-related failure on 32-bit systems

rossbar reviewed (commented):
Instead of removing/reorganizing the "Setting Up" article, I'd be in favor of linking to numpy.org/install from `doc/source/user/install.rst`.

melissawm committed:
BUG: Fixes module data docstrings.

Fixes gh-15325

bjnath commented:
@rossbar  Just to be clear -- there isn't a "Setting Up" article; there's a page with three links. Are you saying we should keep that page with those three links? 

I'll explain why I believe that section needs to go, in hopes that you can explain further why you think it should stay.

* The first article of "Setting Up" is "What is NumPy," which has nothing to do with setting up.

* We're not losing "Setting Up" information. Users are offered help getting started on the first page of numpy.org:
![image](https://user-images.githubusercontent.com/6691888/94094426-762da200-fded-11ea-9055-b9c8aa58764a.png) 

*  Unfortunately, nobody with installation problems will think to check a page titled "Troubleshooting ImportError." The title says nothing about installation, and `ImportError` isn't in the error message. On the other hand, the page becomes discoverable and useful if it's referenced from the install instructions; [PR 360 in the numpy.org repo](https://github.com/numpy/numpy.org/pull/360) adds a sample of the error message and a link to the page.

mattip committed:
Merge pull request #17299 from seberg/maint-dtype-transfer-1

MAINT: (dtype-transfer) make copyswapn and legacy cast wrapper method

seberg committed:
TST: Add test for pickling using ufunc name as qualname

bjnath commented:
@rossbar Also, to make it clear, the "Troubleshooting ImportError" page isn't going away.  Not only is it linked to from the install instructions, where it will be found by people doing installs, it shows up in search -- as you can verify from the [build artifact](https://15881-908607-gh.circle-artifacts.com/0/doc/build/html/index.html), where it shows up in searches for "importerror" and for "please read this advice" (the error message).  

We want to be sure the page isn't overlooked. But the current link [**is** being overlooked,](https://github.com/numpy/numpy/issues/17363#issuecomment-696520358) and -- as I've explained above -- that's not surprising.   This PR, along with numpy.org PR 360, helps ensure the page gets seen, with the additional benefit of reducing clutter on the docs page.

mattip committed:
Merge pull request #17289 from seberg/ufunc-pickle

MAINT: Simplify ufunc pickling

rossbar commented:
I think I understand the overarching goal of this PR + numpy/numpy.org#360. IMO there is value of retaining the structure of this section of the reference docs for linking & discoverability. For example, I don't think that removing the troubleshooting doc from all toctrees increases its visibility (though I do agree it should be linked from numpy.org as well!) I'm for getting rid of outdated info (e.g. replacing the link in `doc/source/user/install` with one to `numpy.org/install`) but I do think there is value in having some intro/setup information in the reference docs in addition to what's on numpy.org.

bjnath commented:
@rossbar 

> I think I understand the overarching goal 

Good, so let's talk details.  I've taken time to explain several drawbacks of the current layout.  

* Why does "What is NumPy" belong in Setting Up?
* Who will read the "Troubleshooting ImportError" section when it has no apparent relation to installation problems?
* Why do we need to lead the docs page with Setting Up when it's prominent in numpy.org?

Comments?

> I don't think that removing the troubleshooting doc from all toctrees increases its visibility 

Nobody has claimed that. 

> intro/setup information 

I'm confused about what "intro/setup information" is.  I'm talking about installation docs. Nobody doubts there's a place in the docs for introductory material.

>  in the reference docs 

We're not talking about the reference docs. We're talking about page 1, line 1, of all the Sphinx documentation. It's an opportunity to do something useful.

> I do think there is value in having some...setup information

I don't. Can you help me understand what the value is?

melissawm committed:
Fixing tests for pypy and win.

melissawm committed:
Adding PR review comment.

albertvillanova committed:
DOC: random: Fix default_rng docstring (#17375)

seberg committed:
Merge pull request #17350 from gerrymanoim/tag-prs-on-prefix

ENH: Add PR prefix labeler and numpy prefix mapping

melissawm committed:
Removing extra comment.

bjnath committed:
DOC: Revise 'contiguous' in PR #16996

bjnath committed:
DOC: Bash examples in PR #17353 now Python

And other suggestions by @eric-wieser.
Also standardizes quotes as '"'.

bjnath committed:
DOC: Italics sted monospace in PR #16996

bjnath committed:
DOC: One more bash->python in PR #17353

melissawm committed:
Added textwrap.dedent to test.

bytesandwich committed:
Minor grammatical correction in quickstart doc.

Single sentence grammatical correction.

mattip committed:
Merge pull request #17380 from bytesandwich/patch-1

Minor grammatical correction in quickstart doc.

bjnath committed:
DOC: NumPy restyling for pydata theme

charris committed:
Merge pull request #17373 from melissawm/f2py-print

BUG: Fixes module data docstrings.

seberg committed:
NEP: Edit nep-0042 for more clarity (#17370)

* DOC: Edit NEP-42 for clarity

* DOC: Further edits of NEP 42

* DOC: Edit NEP-42 for clarity

* DOC: Further edits of NEP 42

* WIP: Work out structure a bit and make intro longer

* DOC: Round 3 of NEP 42 edits

* Start with a new motivation/overview to be forced upon the rest

* restructure along "top-level" categories and rename adjust_descriptors

* Small fixups

* Small clarity/correctness edits

* Use `PyArrayMethod` instead of UFuncImpl

I am starting to feel that calling this "array method" (even if
it is in no way numpy array specific) is probably the best name.
It is not ufunc specific, because ufuncs are dispatching + array method.
Calling it ufuncimpl makes sense, but since the UFunc name which would
also work perfectly is already taken, maybe a whole new name is better.

* DOC: Fold in @seberg revisions

* Fix typo

* Make first example/overview inherit dtype instead of metaclass

it could say `metaclass=DTypeMeta`, but its implicit.

* DOC: Add "checklist" to NEP 42

Also anchor NEP 40 figure for reference in NEP 42.

* Delete duplicated dtype-discovery section

The section must have been duplicated in an merge conflict :/,
I scanned through, and I think the version that was in the right place
got most edits, although not 100% sure.

* DOC: Reorganize scope, standardize heds

* DOC: Retitle, revisions up to Casting

Also anchor for parametric type discussion in NEP 40

* Some smaller changes/additions

* DOC: Reorganize for better flow

Also adds a NEP anchor at top.

* DOC: Glossary, shuffle "Additional information" bullets

* DOC: Small updates for sense and case

* Fixup merge (some changes were not saved when finishing merge)

* Answer bjnath's questions and motivate user impact

It seems most answers bjnath already included himself, this adds
a few tweaks.
Added a note that tusers will basically have to implement the
DType class as outlined is basically what the user will have to
implement, so spell that out both at the top and in the C-API to
clarify that.

* NEP 41: Add subsection on dtype organization with parameters

This might go a bit far, but I think the image goes very far in
giving people the right idea of how to think of dtypes and
DType instances.

The question is if we want to even show the "alternative"...

* Add bjnath as author

* Add new header also to NEP 41

* DOC: Minor: delete horiz rules, tweak 2 headings

* Rephrase NEP 41 bullt points in terms of NEP 42

Also changes those for NEP 43, to represent what is done, rather
than what is solved (and thus fit better to the way its written
for the NEP 42 part)

* Try to rewerd paragraphs around setitem/getitem a bit

some of these were very confusing, hopefully they are a bit better
now.

* Make a pass over the array-coercion part

* Make careful pass up until array-coercion

* Pass over the second half of the NEP 42 draft.

* Use "given_descrs" which seems clearer and fix casting

I always thought returning the casting safety is an awesome idea, but
right now I am actually not so sure it helps much with anything.
Its usually simpler, I guess. But if the user requested safe casting
and we have an unsafe cast, for example for structured dtypes we
will end up doing a lot of work for no reason.

If we accept an error return without an error already set this is
solved. My original reason was that setting a generic error in the
resolve descriptor function would really be pretty annoying, so
that returning the casting safety seemed like a good way to
avoid worrying about most errors there.

The "cast-is-view" output is important though.

* Fix copy new intro note to NEP 40 (and fix it in NEP 41)

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.35.3 to 5.36.1

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.35.3 to 5.36.1.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.35.3...hypothesis-python-5.36.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

charris committed:
Merge pull request #17386 from numpy/dependabot/pip/hypothesis-5.36.1

MAINT: Bump hypothesis from 5.35.3 to 5.36.1

charris committed:
MAINT: Finish replacing PyInt_Check (#17364)

* MAINT: Replace PyInt_Check in multiarray_tests.c.src

* MAINT: Replace PyInt_Check in number.c

* MAINT: Replace PyInt_Check in interators.c

* MAINT: Replace PyInt_Check in refcount.c

* MAINT: Replace PyInt_Check in hashdescr.c

* MAINT: Replace PyInt_Check in scalarapi.c.

* MAINT: Replace PyInt_Check in scalartypes.c.src.

charris committed:
MAINT: Remove old debug print statement.

This looks like an old debug statement that should have been removed
long ago. The C stdio.h header is nowhere imported and PyInt_Type no
longer exists in Python 3.

seberg committed:
Merge pull request #17388 from charris/cleanup-debug-print

MAINT: Remove old debug print statement.

charris committed:
ENH: Add metastr_to_unicode function to datetime.c.

This new function is intended to be used instead the existing
append_metastr_to_string function. It returns the datetime metastr as a
unicode string rather than appending it to an existing unicode string.
Uses of append_metastr_to_string are replaced in datetime.c.

jftsang committed:
DOC: Fix docstring for np.matmul (#17383)

Update argument names to match ufunc signature in matmul docstring.

bjnath committed:
DOC: Replace "About NumPy" with "Document conventions"

Eliminates the unhelpful "About NumPy" title. All contents
of the original appear more prominently elsewhere,
except for the information about document conventions,
which is now the title and subject.

mattip committed:
Merge pull request #16996 from bjnath/revised-glossary

DOC: Revise glossary page

bjnath committed:
DOC: Update info on doc style rules

charris committed:
MAINT: Replace append_metastr_to_string in scalartypes.c.src

Use metastr_to_unicode instead.

seberg committed:
BUG: Fix default void, datetime, and timedelta in array coercion

When converting an empty sequence to an array, datetimes would
use an incorrect itemsize (on master only).
The legacy behaviour of void is that it uses 8 bytes, which is
due to the fact that `[]` uses `float64` by default which has
8 bytes.

charris committed:
MAINT: Replace append_metastr_to_string in descriptor.c.

Use metastr_to_unicode instead.

charris committed:
Merge pull request #17393 from seberg/fix-default-parametric-dtypes

BUG: Fix default void, datetime, and timedelta in array coercion

charris committed:
MAINT: Revise comment in numpy.core._dtype.py

Replace append_metastr_to_string by metastr_to_unicode.

charris committed:
MAINT: Remove append_metastr_to_string function.

It has been replaced by metastr_to_unicode.

mattip committed:
Merge pull request #17392 from bjnath/style-guides

DOC: Update info on doc style rules

leakec committed:
BLD: Fixed ARGOUTVIEWM memory deallocation.
Closes #17398

charris committed:
MAINT: Replace PyUString_ConcatAndDel in mapping.c.

eric-wieser committed:
Merge pull request #17396 from charris/cleanup-append_metastr_to_string

MAINT: Replace append_metastr_to_string function.

charris committed:
MAINT: Replace PyUString_ConcatAndDel in nditer_constr.c.

rossbar committed:
DOC: rm incorrect alias from recarray user article.

melissawm commented:
I'm in favor of deduplication of content and better organizing this information, and I agree the current Install page can be removed. But I think there's more to it.
- The [Building from source](https://numpy.org/devdocs/user/building.html) only shows up in the left sidebar once a user clicks on any of the User Guide subpages, or in the Development Environment page. I think this should be discoverable somewhere in the user guide, but I'm not sure if all agree. 
- What about [INSTALL.rst.txt](https://github.com/numpy/numpy/blob/master/INSTALL.rst.txt)? This contains a lot of information that is also at the `Building from source` page. Should we have duplicate instructions for that? (this can be dealt with in a separate PR but if we're going to delete the Install page I think we need to think about this too)

rossbar committed:
DOC: Add arraysetops to autosummary.

 * Generate stub for module, fixes 7 broken links
 * Modify arraysetops module docstring w/ rst formatting
 * Remove listing of functions from module docstring.

rossbar committed:
Incorporate feedback from review.

mattip commented:
> What about `INSTALL.rst.txt`? This contains a lot of information that is also at the "Building from source page"

The `INSTALL.rst.txt` page historically was linked to directly from other places like SciPy, but I see they only have one [weak reference](https://github.com/scipy/scipy/blob/95704c3bda7ce1d2623e603ea19073139579f291/scipy/linalg/tests/test_build.py#L54) that could easily be changed (since they have [their own copy](https://github.com/scipy/scipy/blob/master/INSTALL.rst.txt)). It has an internal reference in [doc/HOWTO_RELEASE.rst.txt](https://github.com/numpy/numpy/blame/19f91f3935069229bbc23bbda4294c36957901a5/doc/HOWTO_RELEASE.rst.txt#L14)

mattip committed:
Merge pull request #17402 from rossbar/doc/add_arysetops

DOC: Add arraysetops to an autosummary

rgommers commented:
If we could leave the file `INSTALL.txt` in place, with it containing just a link to where the actual info lives, that'd be fine with me. We shouldn't remove the file itself, because this is what people will look for when all they have is a source release or repo checkout.

rgommers committed:
Merge pull request #17382 from bjnath/new-pydata-css

DOC: NumPy restyling for pydata theme

bjnath committed:
DOC: Updated install.rst in PR #17365

Per @rgommers, @eric-wieser. Just has a link to the numpy.org install
page.

bjnath commented:
@rgommers 
cc @eric-wieser 

> If we could leave the file INSTALL.txt in place

ebb96be restores `install.rst` (I assume this is what you meant) with a link to the Hugo instructions, so `user/install.html` will still exist.   Did you also want it exposed in the toc?

mattip committed:
Merge pull request #17399 from leakec/argoutviewm_array-fixes

BLD: Fixed ARGOUTVIEWM memory deallocation. Closes #17398.

charris committed:
Merge pull request #17400 from rossbar/doc/correct_rec_alias

DOC: rm incorrect alias from recarray user article.

Commit:
ENH: Replace the module-level `__getattr__` with explicit type annotations #17367

seberg committed:
API: Special case how numpy scalars are coerced to signed integer

This removes one of the larger changes to array-coercion, which
meant that NumPy scalars were always coerced like a 0-D array
would be (i.e. using normal casting). When the assignment is
explicitly an integer, now `scalar.__int__()` will be used instead
(as was the case previously).
Since previously this was handled differently, a *single* scalar
is still converted using casting:

    np.array(np.float64(np.nan), dtype=np.int64)

succeeds, but any other thing fails, such as:

    np.array([np.float64(np.nan)], dtype=np.int64)
    arr1d_int64[()] = np.float64(np.nan)
    np.array(np.array(np.nan), dtype=np.int64)

This does not affect Python scalars, that always raise, because
they always are converted using `scalar.__int__()`.

Unsigned integers always supported casting from their signed
equivalent, so the difference is much less visible for them and
this chooses to always use the casting behaviour.

The main reason for this change is to help pands:
https://github.com/pandas-dev/pandas/issues/35481

Commit:
ENH: remove unnecessary annotations in __init__.pyi.

Commit:
TST: Add test to verify that getting a non-existent objects indeed results in an error.

rgommers committed:
DOC: in PR template, set expectations for PR review timeline

mattip committed:
Merge pull request #17407 from rgommers/pr-template

DOC: in PR template, set expectations for PR review timeline

charris committed:
MAINT: Cleanup PyUString_ConcatAndDel in scalartypes.c.src.

charris committed:
MAINT: Replace PyUString_ConcatAndDel in shape.c.

charris committed:
MAINT: Replace PyUString_ConcatAndDel in array_assign.c

Commit:
TST: Added tests for bitwise operations

charris committed:
MAINT: Remove the build_shape_string function.

The build_shape_string function is effectively a duplicate of the
convert_shape_to_string function in common.c. All uses have now been
replaced, so remove it.

charris committed:
MAINT: Replace PyUString_ConcatAndDel in common.c.

charris committed:
MAINT: Replace PyUString_ConcatAndDel in descriptor.c.

charris committed:
MAINT: Replace PyUString_ConcatAndDel in ufunc_object.c.

Commit:
Revert "TST: Add test to verify that getting a non-existent objects indeed results in an error."

This reverts commit 0996962236506342d16730580e126efe2dfbf98d.

Commit:
TST: add test to verify that getting a non-existent objects of top-level indeed results in an error

mattip committed:
Merge pull request #17406 from junjunjunk/is17367

ENH: Replace the module-level `__getattr__` with explicit type annotations #17367

Commit:
TST: Mark the typing tests as slow

charris committed:
Merge pull request #17411 from BvB93/mypy-error

TST: Mark the typing tests as slow

Commit:
MAINT: Removed placeholder annotations

eric-wieser committed:
Merge pull request #17405 from charris/cleanup-PyUString-in-mapping.c

MAINT: Replace PyUString_ConcatAndDel in mapping.c.

charris committed:
Merge pull request #17326 from BvB93/asarray

ENH: Add annotations for five array construction functions

BvB93 committed:
DOC: Fix a parameter type in the `putmask` docs (#17412)

Commit:
MAINT: Move the `np.core.numeric` functions to their own stub file

melissawm commented:
> Are you suggesting we add an `Installing from source` link to the docs landing page (numpy.org/devdocs)? If so, maybe it's better to make the link `Installation` and have it go directly to the Hugo page.

That sounds fine to me. I'm imagining a workflow like: land on numpy.org/doc/stable (where devdocs will end up in the next release) and looking for a link to installation instructions. I'd have no problem if the link in this toc is directly to the hugo page, but I do think it has to be somewhere.

seberg committed:
DEP: Deprecate coercion to subarray dtypes

When coercing to subarray dtypes, e.g. using `np.array(obj, dtype)`,
but also `arr.astype(dtype)`, the behaviour was only well defined
with tuple inputs, but not with array-like inputs.
In particular, `arr.astype(dtype)` had arguably surprising behaviour
of not converting by element, but rather attempting (and often failing)
to broadcast `arr` to the result array with added dimensions.

This deprecates all of these cases, the main issue would be for users
relying on stranger inputs with broadcasted tuples contained in
sequences:
```
np.array([((0, 1), (1, 2)), ((2,),)], dtype='(2,2)f4')
```

In most cases, where the tuples have the correct output shape,
the new base dtype can be directly used since the discovered shape
should match.
However, there is no work-around for the above case.

Closes gh-17173

seberg committed:
BUG: Fix memory leak in array-coercion error paths

Also uses PyMem_*, since those seem the more approriate functions.

nandgator committed:
fix: chains nested try-except-raise

quique committed:
DOC: Remove bogus reference to _a_

This will make this docstring coherent with the one for ndarray.argmax().

eric-wieser committed:
Merge pull request #17423 from quique/fixes

DOC: Remove bogus reference to _a_

eric-wieser committed:
DOC: Fix formatting issues in description of .c.src files

eric-wieser committed:
Update doc/DISTUTILS.rst.txt

charris committed:
Merge pull request #17424 from eric-wieser/_c.src-docs

DOC: Fix formatting issues in description of .c.src files

mattip committed:
Merge pull request #17422 from joe733/hotfix-for-nested-try

MAINT: chains nested try-except in numpy/ma/core.py

charris committed:
Merge pull request #17409 from charris/cleanup-remaining-PyUString-use

MAINT: Cleanup remaining PyUString_ConcatAndDel use.

charris committed:
Merge pull request #17404 from charris/cleanup-PyUString-in-nditer_constr.c

MAINT: Replace PyUString_ConcatAndDel in nditer_constr.c.

BvB93 committed:
ENH: Annotate the arithmetic operations of `ndarray` and `generic` (#17273)

* ENH: Added annotations for arithmetic-based magic methods

* TST: Added arithmetic tests

* TST: Moved a number of tests to `arithmetic.py`

* ENH: Ensure that objects annotated as `number` support arithmetic operations

* MAINT: Arithmetic operations on 0d arrays return scalars

* MAINT: Clarify the type of generics returned by `ufunc.__call__`

* TST: Added more arithmetic tests

* MAINT: Use `_CharLike` when both `str` and `bytes` are accepted

* MAINT: Change the `timedelta64` baseclass to `generic`

* MAINT: Add aliases for common scalar unions

* MAINT: Update the defition of `_NumberLike`

* MAINT: Replace `_NumberLike` with `_ComplexLike` in the `complexfloating` annotations

* MAINT: Move the callback protocols to their own module

* MAINT: Make `typing._callback` available at runtime

* DOC: Provide further clarification about callback protocols

* MAINT: Replace `_callback` with `_callable`

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r485821346

The use of `__call__`-defining protocols is not limited to callbacks. The module name name & docstring now reflects this.

* MAINT: Removed `__add__` from `str_` and `bytes_`

Most `np.bytes_` / `np.str_` methods return their builtin `bytes` / `str` counterpart.
This includes addition.

* MAINT: Fix the return type of boolean division

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r486271220

Dividing a `np.bool_` by an integer (or vice versa) always returns `float64`

* MAINT: Renamed all `_<x>Arithmetic` protocols to `_<x>Op

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r486272745

* TST: Add tests for boolean division

* ENH: Make `np.number` generic w.r.t. its precision

* ENH,WIP: Add a mypy plugin for casting `np.number` instances to appropiate subclasses

* Revert "ENH,WIP: Add a mypy plugin for casting `np.number` instances to appropiate subclasses"

This reverts commit c526fb619d20902bfd77709c8983c7a7d5477c95.

* Revert "ENH: Make `np.number` generic w.r.t. its precision"

This reverts commit dbf20183cf7ff71e379cd1a165d07e1a1d643135.

* MAINT: Narow the definition of `_ComplexLike`

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r490440238

* MAINT: Refined the return type of `unint + int` ops

`unsignedinteger + signedinteger` generally returns a `signedinteger` subclass.
The exception to this is `uint64 + signedinteger`, which returns `float64`.

Addresses https://github.com/numpy/numpy/pull/17273#discussion_r490442023

* MAINT: Use `_IntLike` and `_FloatLike` in the definition of `_ComplexLike`

chsafouane committed:
nep-0029 typo correction

to provides -> to provide

seberg committed:
BUG: Fix memoryleaks related to NEP 37 function overrides

This adds a few missing DECREFs/XDECREFs, all of these are already
covered by the tests and can be found running a leak-checker.

charris committed:
Merge pull request #17430 from seberg/fix-memory-leak-nep37

BUG: Fix memoryleaks related to NEP 37 function overrides

takanori-pskq committed:
DOC: Fix the links for ``Ellipsis`` (#17431)

arogozhnikov committed:
DOC: add references to einops and opt_einsum

Following discussion in mailing list, this PR
adds mentions to third-party libs einops and opt_einsum
to einsum documentation

charris committed:
Merge pull request #17427 from chsafouane/patch-1

NEP: nep-0029 typo correction

charris committed:
Merge pull request #17421 from seberg/coercion-cache-leak-large-arrays

BUG: Fix memory leak in array-coercion error paths

charris committed:
MAINT : Disable 32 bit PyPy CI testing on Windows.

The test is consistently failing at the moment, we can enable it again
when PyPy gets fixed.

bjnath committed:
DOC: Security warning for issues template

Tag for "Bug report" now warns not to file security vulnerabilities.

Fixes gh-17434.

bjnath committed:
DOC: Fix "Feature request" spelling in issue templates

"Request" had an upper-case R, inconsistent with Google style and
other template items.

charris committed:
Merge pull request #17433 from charris/disable-pypy-test-on-windows

MAINT : Disable 32 bit PyPy CI testing on Windows.

mattip committed:
Merge pull request #17435 from bjnath/security-issue-template

DOC: Security warning for issues template

mattip committed:
Merge pull request #17436 from bjnath/issue-template-spelling

DOC: Fix "Feature request" spelling in issue templates

bjnath committed:
DOC: Revise credit line per CC

Adds "CC BY 4.0" credit to the Pauli Vertanen example.

devhpes committed:
fix exception chaining in format.py

bjnath committed:
DOC: Cleaner PR template

Make the instructions easy to see and hard to delete unread.

Moves some text to new sections in dev/development_workflow.rst.

The dashes above and below the heading, which
in monospace seem too long, fit it
well in Chromium browsers and reasonably in Firefox.

bjnath committed:
DOC: Fixes for PR 17440

Colon missing in heading.

Added sphinx anchor for PR-timeline (html anchor was OK).

bjnath committed:
DOC: Fix indents in PE #17440

hugovk committed:
ENH: Warn on unsupported Python 3.9+

hugovk committed:
MAINT: Fix Flake8

hugovk committed:
Add comment to remind to keep consistent with classifiers

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

hugovk committed:
Clarify warning, it might work before official support

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

hugovk committed:
Already compatible with 3.10

hugovk committed:
Update comment

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

hugovk committed:
Add 3.10 classifier

takanori-pskq committed:
DOC: Fix the references for macros

Commit:
ENH: Add  `Typing :: Typed` to the PyPi classifier

charris committed:
Merge pull request #17443 from hugovk/warn-3.9-plus

ENH: Warn on unsupported Python 3.10+

bjnath committed:
DOC: Remove CoC pages from Sphinx

code_of_conduct.rst and report_handling_manual.rst duplicate
pages now on Hugo.

mattip committed:
Merge pull request #17444 from BvB93/typed

ENH: Add `Typing :: Typed` to the PyPi classifier

eric-wieser committed:
DOC: Improve [source] links for C extension types

Previously, these would all link to `numpy/core/__init__.py`.
Now the scalar type and `ndarray` link to the files where the `PyTypeObject` is defined.
In future, we should do this for all extension types, probably automatically.

charris committed:
Merge pull request #17440 from bjnath/bolder-PR-template

DOC: Cleaner template for PRs

charris committed:
Merge pull request #17432 from arogozhnikov/add-einops-opt_einsum-reference

DOC: add references to einops and opt_einsum

eric-wieser committed:
DOC: Change the value of type.__name__ on the scalar types for sphinx builds

By default, the `.__name__` of the numeric `np.generic` subclasses is their bitlength name, such as `np.int64`.
This is convenient when working interactively, because it lets users see the size of their array easily; but in docs it is confusing, as the sizes of the integers in the doc build may not match their size on the platform of the user reading them.
Without this change, `..autoclass:: numpy.short` would just display "alias of uint16", which is backwards.

Rather than changing the names globally, or adding a build flag to change the names, this uses `ctypes` to modify the scalar names at startup.
This resembles the approach taken by the `forbiddenfruit` module for patching builtin slots, although that would be overkill here.

The timing of when we perform this patching is important - we can't do it until after `numpy.core._umath_multiarray` has been loaded, but we need to do it before `numpy.core._add_newdocs` generates the name-based docstrings.
Similarly, we can't just disable `numpy.core._add_newdocs` until later, as it populates docstrings in `ndarray` on which `numpy.ma.core` does further processing.
To resolve this, we split out the scalar docstrings in `numpy.core._add_newdocs` into a new module `numpy.core._add_newdocs_scalars` that _is_ safe to disable until later.

eric-wieser committed:
DOC: Overhaul the scalar type documentation to use `..autoclass`

This remove the tables. since they only had three columns, and using the character code is advised against anyway.

With this change, the individual scalar types as well as their aliases are now valid sphinx python domain targets.

charris committed:
Merge pull request #17448 from bjnath/rm-dupe-coc

DOC: Remove CoC pages from Sphinx

takanori-pskq committed:
Fix: Add the entry for `NPY_ARR_HAS_DESCR`

takanori-pskq committed:
DOC: Fix some references for macros

mattip committed:
Merge pull request #17445 from takanori-pskq/i13114-fix-refs-macro

DOC: Fix the references for macros

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.36.1 to 5.37.0

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.36.1 to 5.37.0.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.36.1...hypothesis-python-5.37.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

stuartarchibald committed:
DOC: update code of conduct URL

As title.

Closes: #17458

charris committed:
Merge pull request #17455 from numpy/dependabot/pip/hypothesis-5.37.0

MAINT: Bump hypothesis from 5.36.1 to 5.37.0

charris committed:
Merge pull request #17372 from BvB93/from-numeric

ENH: Add annotations for remaining `ndarray` / `generic` non-magic methods

mattip committed:
Merge pull request #17459 from stuartarchibald/fix/17458

DOC: update code of conduct URL

stuartarchibald committed:
BUG: Fixes incorrect error message in numpy.ediff1d (#17457)

Fixes the error message reported by ediff1d in the case of
an invalid to_begin and updates tests to assert validity.

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

rossbar committed:
DOC: typo.

charris committed:
Merge pull request #17391 from bjnath/doc-conventions

DOC: Replace "About NumPy" with "Document conventions"

charris committed:
Merge pull request #17353 from bjnath/how-to-io

DOC: How-to guide for I/O

takanori-pskq committed:
DOC: Add some entries for C types and macros

mattip committed:
Merge pull request #17331 from eric-wieser/remove-scalar-tables

DOC: Remove the tables of scalar types, and use `..autoclass` to create link targets instead

takanori-pskq committed:
Fixup

takanori-pskq committed:
Fix: Remove the link for `PyArrayMapIterObject`

mattip committed:
Merge pull request #17464 from takanori-pskq/i13114-add-entries

DOC: Add some entries for C types and macros

Commit:
REV: Remove the `uint64`/`signedinteger` specific overload

Unfortunately the likes of `self: uint64` don't work with protocols.

Revisit this once we add proper support for numerical precision.

eric-wieser committed:
DOC: add some missing scalar aliases

eric-wieser committed:
DOC: Fix broken references

Unfortunately we can't use `np.` as a module name alias

jsignell committed:
TST: Fix doctest for full_like

rossbar committed:
DOC: rm numpy. from type alias attr definitions.

person142 committed:
NEP: update NEP 42 with discussion of type hinting applications (#17447)

* NEP: update NEP 42 with discussion of type hinting applications

As discussed in https://github.com/numpy/numpy/pull/16759, the new
DType classes provide a good path forward for making `ndarray` generic
over DType. Update NEP 42 to discuss those applications in more
detail.

* NEP: discuss typing for use scalar types in NEP 42

Also clean up the language a bit.

Commit:
MAINT: Add an empty `__init__.pyi` file to `np.core`

Prevents mypy crashes related to import cycles.

See https://github.com/numpy/numpy/issues/17316

charris committed:
Merge pull request #17472 from jsignell/patch-2

TST: Fix doctest for full_like

takanori-pskq committed:
DOC: Fix the entries for members of structures

valdaarhun committed:
MAINT: Chain exceptions in "_polybase.py" (#17453)

* Removed "from" keyword and changed "raise e" to "raise"

mattip committed:
Merge pull request #17474 from BvB93/numeric

MAINT: Move the `np.core.numeric` annotations to their own stub file

person142 committed:
ENH: type np.unicode_ as np.str_

The two are aliases; i.e.

>>> np.unicode_ is np.str_
True

so make them aliases on the typing level too.

mattip committed:
BLD: circleCI- merge before build, add -n to sphinx

mattip committed:
Update .circleci/config.yml

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

mattip committed:
Merge pull request #17468 from eric-wieser/remove-scalar-tables

DOC: add some missing scalar aliases

mattip committed:
Merge pull request #17442 from devhpes/devhpes

MAINT: fix exception chaining in format.py

mattip committed:
BLD: try merging in a separate step

Commit:
MAINT: Removed a number of duplicate placeholders

mattip committed:
BLD: don't use -q in sphinx build

Commit:
MAINT: Ignore three builtin modules present in the numpy namespace

Commit:
TST: Updated the module-related tests

Commit:
MAINT: Ignore the private `np.compat` module

Ashutosh619-sudo committed:
MAINT: Chaining exception in numpy\numpy\ma\mrecords.py (#17438)

* fix runF2c exception

mattip committed:
Merge pull request #17481 from takanori-pskq/i13114-fix-structures-member

DOC: Fix the entries for members of structures

BvB93 committed:
MAINT: Move aliases for common scalar unions to `numpy.typing` (#17429)

* MAINT: Move the `<scalar>Like` unions to `numpy.typing`

takanori-pskq committed:
DOC: Fixing references in ma docstrings (#16300)

Fix broken links in see also section of some ma functions.

mattip committed:
Merge pull request #17320 from seberg/relax-object-dtype-with-ref

BUG: allow registration of hard-coded structured dtypes

mattip committed:
Merge pull request #17410 from seberg/scalars-to-int-array-special

API: Special case how numpy scalars are coerced to signed integer

mattip committed:
Merge pull request #17419 from seberg/deprecate-coercion-to-subarray-dtype

DEP: Deprecate coercion to subarray dtypes

bjnath committed:
DOC: Use consistent lowercase on docs landing page

Corrects titles to all use sentence case, per style guide.

seberg committed:
BUG: Fix usertype typenumber during registration

This slightly rearranges the code to ensure that the type number
set for a user type is -1 (invalid) if it does not succeed.
(This should be unnecessary, but might be relevant in the unlikely
event that a usertype does not check the error return correctly.)

seberg committed:
BUG: Accept legacy user-dtypes if not isinstance(dtype.type, np.generic)

These are now accepted (as a no-op) in `_PyArray_MapPyTypeToDType`
retaining previous behaviour (as lookup for a registered user-type
was only done if the instance check suceeded.

Retain the check for future user DTypes, as it sould be a conscience
choice to relax this requirement.

person142 committed:
MAINT: add str0 as an alias of str_ on the typing level

person142 committed:
BUG: remove `sys` from the type stubs

It is the builtin sys module, which people should not be accessing
through the NumPy namespace.

bjnath committed:
DOC: New round of NEP 42 edits (gh-17493)

* DOC: Revise NEP 42 "common DType" section

* DOC: Eliminate duplicate doc in NEP42

* DOC: Fix missing 'not' in NEP42 edit

* DOC: Minor typos in NEP42 edit

* DOC: Reword opening sentence of NEP 42 revision

Also clarify a wording.

* DOC: Correct :class: reference in NEP42 edit

* DOC: Fix Windows line-enders on NEP42 edit

* DOC: New round of common DType edits of NEP 42

* DOC: Uniquify anchors in NEP 42

* DOC: Typo in NEP42 edit

* DOC: Last push for NEP 42 edit

* Small fixups and moving example

One paragraph on `__common_dtype` in the alternatives was already
before very confusing. This makes it hopefully more clear.

* NEP: Change title to "New and extensible DTypes"

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

mattip committed:
Merge pull request #17501 from seberg/fix-master-failure

BUG: Fix failures in master related to userdtype registeration

mattip committed:
Merge pull request #17497 from bjnath/lowercase-mainpage

DOC: Use consistent lowercase on docs landing page

rgommers committed:
Merge pull request #17502 from person142/stubs-sys

BUG: remove `sys` from the type stubs

roysti10 committed:
Update __init__.pyi

roysti10 committed:
ENH: Marked all constants as `Final

roysti10 committed:
Minor bug fix

roysti10 committed:
Adjusted typing.Final's python 3.8 compatibility

bjnath committed:
DOC: Fewer blank lines in PR template

Reduce the need to scroll, per @mattip suggestion in #17440

mattip committed:
Merge pull request #17510 from bjnath/squeeze-pr-template

DOC: Fewer blank lines in PR template

bjnath committed:
DOC: Rename 'Quickstart tutorial' (#17504)

Remove tutorial wording in favor of article, minor wording updates.

takanori-pskq committed:
DOC: Fix random links in release note (#17483)

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris committed:
Merge pull request #17487 from BvB93/duplicate

MAINT: Remove duplicate placeholder annotations

charris committed:
Merge pull request #17485 from mattip/circleci1

BLD: circleCI- merge before build, add -n to sphinx

roysti10 committed:
[CODE] Made code cleaner

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

roysti10 committed:
[CODE]Redundant spaces removed

Co-authored-by: Bas van Beek <43369155+BvB93@users.noreply.github.com>

charris committed:
Merge pull request #17465 from BvB93/bitwise-ops

ENH: Add annotations for bitwise operations

charris committed:
Merge pull request #17479 from person142/np-unicode

ENH: type np.unicode_ as np.str_

charris committed:
Merge pull request #17508 from lucasace/typing-final-feature

ENH: Added the Final feature for all constants

charris commented:
@bjnath Needs rebase.

bjnath committed:
DOC: Rebase PR #17365; add Installation link per @melissawm

bjnath committed:
Merge branch 'delete-install-page' of github.com:bjnath/numpy-1 into delete-install-page

# Conflicts:
#	doc/source/_templates/indexcontent.html

bjnath commented:
Rebase attempt went bad, closing.

charris commented:
Oops :)  Just do `git rebase master` next time and fix the conflicts. It helps to squash the commits if you have a bunch of them. You can also fix this by merging it to a new pr branch from current master, the extra commits will go away.

bjnath commented:
Thanks, @charris -- I saw those hundreds of commits and panicked. Resurrected as #17523.

charris commented:
Slightly more complicated, you can also do a hard reset to the state before the oops. To find the hash for that state do `git reflog`.
